### PR TITLE
Allow for the iteration of ListObjects.

### PR DESCRIPTION
**Pull request by mastermatt**
More pythonic `for .. in` loops can now
loop the the contents of the `data` list
and continue on using the cursor-based
pagination model.

```python
for charge in stripe.Charge.all():
    my_func(charge)
```

### CODE REVIEW

File: stripe/resource.py [hunk 1]
```diff
@@ -316,6 +316,22 @@ def retrieve(self, id, **params):
...
         return self.request('get', url, params)
 
+    def __iter__(self):
+        page = self
+        params = dict(self._retrieve_params)  # copy so not to mutate originals
+
+        while True:
+            last_id = None
+            for item in page.data:
+                last_id = item.get('id')
```
Comment (kyleconroy) [inline stripe/resource.py]: We can't assume that all list objects will have an `id`. To be extra safe, we should use `.get('id', None)`

Comment (kyleconroy) [inline stripe/resource.py]: Whoops, ignore me :)

File: stripe/test/test_resources.py [hunk 2]
```diff
@@ -208,6 +208,60 @@ def test_retrieve(self):
         self.assertResponse(res)
 
 
+class ListObjectTests(StripeApiTestCase):
```
Comment (kyleconroy) [inline stripe/test/test_resources.py]: We'll need to add tests for edge cases. Specifically, we need to cover the case when `has_more` is `True`. We also should test the case when we fetch an additional page and no results are returned.

Comment (mastermatt) [inline stripe/test/test_resources.py]: I didn't bother to go down that road until I received feedback. I wanted to ensure the core devs were ok with the __iter__ recalling the endpoint. It could be reasoned that iterating the list object should only go through the single page.

Comment (kyleconroy) [inline stripe/test/test_resources.py]: Another option would be to add an `iter()` method to list objects, making it clearer that it will return all objects. Thoughts?

**Comment by kyleconroy** [comment]
We may not want to make the pagination on by default. We could do something like this?

```python
stripe.auto_paginate = True

for charge in stripe.Charge.all(limit=3 paginate=True):
    print(charge)
```
